### ISSUE TITLE
Problems with ETAGO's in HTML5 Documents

### ISSUE THREAD

**Issue by bkonia** [FluentDOM/FluentDOM#45]
Since Fluent uses DOMDocument as its HTML parser, it suffers from a limitation of DOMDocument, in that any ETAGO's contained within a SCRIPT tag will prematurely end the script block, causing your script to fail. For example, the following block:

```
<script type="text/template" id="tmpl-variation-template">
    <div class="woocommerce-variation-description">
        {{{ data.variation.variation_description }}}
    </div>
</script>
```

will be transformed by FluentDOM into:

```
<script type="text/template" id="tmpl-variation-template">
    <div class="woocommerce-variation-description">
        {{{ data.variation.variation_description }}}
    </script>
</div>
```

This issue is discussed in detail at the following URL's:
http://stackoverflow.com/questions/4029341/dom-parser-that-allows-html5-style-in-script-tag
https://mathiasbynens.be/notes/etago

I'm wondering if there's any way, you could extend FluentDOM\Document to work around this DOMDocument limitation and handle this properly?

**Comment by ThomasWeinert** [comment]
Have you tried using the HTML5 Loader/Serializer? https://github.com/FluentDOM/HTML5

**Comment by bkonia** [comment]
Hi Thomas,

Thanks for the suggestion, I didn't realize there was an HTML5 loader. Unfortunately, I haven't been able to get the loader to work at all. I installed it via Composer, but FluentDOM seems to be unaware that it's installed. When I load a document specifying the html5 content type, it loads as XML. My composer.json contains the following:

```
{
    "require": {
        "fluentdom/fluentdom": "^5.3",
        "fluentdom/selectors-phpcss": "^1.0",
        "FluentDOM/Selectors-Symfony": "^1.0",
        "fluentdom/html5": "^1.0"
    }
}
```
I've also verified that the vendor/fluentdom/html5 and vendor/masterminds/html5 directories exist.

**Comment by bkonia** [comment]
Any suggestions on how to get the HTML5 loader working? Is it just me, or is the loader not working for other people?

**Comment by ThomasWeinert** [comment]
Sorry I was a little busy. I will try to reproduce the issue the next days. The error message is not what I would expect. It should report that it can not load the source as content type 'text/html5'.

**Comment by ThomasWeinert** [comment]
I tried to reproduce the issue, but it works for me. Here is my test file (`src/load-file.php`):

```php
require '../vendor/autoload.php';

$fd = FluentDOM('test.html', 'text/html5');
$html5 = new FluentDOM\HTML5\Serializer($fd->document);
echo $html5;
```

Be aware that the DOM will be XHTML (including the namespace). The HTML5 parser converts the input into XHTML so that DOM can work with it. The serializer will convert the XHTML back to HTML5.

I used your composer.json. Here is on small mistake but it has no effect. Installing two CSS selector packages makes no sense - only one will be registered and used. You can even reduce it to:

```json
{
  "require": {
    "fluentdom/selectors-phpcss": "^1.0",
    "fluentdom/html5": "^1.0"
  }
}
```
`FluentDOM/FluentDOM` is a dependency of the two other packages.

**Comment by bkonia** [comment]
Well, it sort of works, but not really. I was able to load the document using the HTML5 loader, and to output as HTML5 using the serializer, but many of my CSS selectors stopped working on the parsed document. Elements I was previously able to select and transform became unselectable and some of my transformed pages failed to render in the browser. I'm assuming this is happening because when the document is converted to XHTML, something changes in the structure that messes up the parser.

If you're interested, I could send you the HTML source for one of the pages, along with the PHP code I'm using to transform the document.

**Comment by ThomasWeinert** [comment]
Did you register and use the XHTML namespace? By default the HTML5 parser creates XHTML with the proper namespaces (XHTML, SVG, MathML, ...). That means you will need to register and use a prefix, or use the universal selector:

```php
$fd = FluentDOM::QueryCss('test.html', 'text/html5');
$fd->registerNamespace('x', 'http://www.w3.org/1999/xhtml');
$html5 = new FluentDOM\HTML5\Serializer($fd->document);

// not found
var_dump($fd->find('script')->text());
// found
var_dump($fd->find('x|script')->text());
var_dump($fd->find('*|script')->text());
```

**Comment by bkonia** [comment]
That helped a lot; I'm now able to select elements. However, there are still several issues:

1. The Universal Selector does not work at all. For example:

```
$fd->find('x|div.someclass') works

$fd->find('*|div.someclass') does not work

$fd->find('x|*.someclass') does not work

$fd->find('x|.someclass') does not work
```

2. Certain documents are crashing (white screen) when I load them using the HTML5 parser. I'm not sure what's causing the crashes, but I presume either something in the document is causing the parser to fail during the XHTML conversion, or it's creating invalid XHTML, which is crashing DOMDocument. I could send you an example of a document that crashes, if you want.

**Comment by ThomasWeinert** [comment]
Yes, the CSS to Xpath transformation library matters. But PHPCss supports the attribute contains selector:

http://xpath.thomas.weinert.info/?selector=div%5Bclass*%3D%22test%22%5D&options%5B1%5D=1

It seems that Symfony CSS Selectors do not support the universal selector for namespaces.

The CSS Selector `*|script` transforms to:
* in PhpCSS: `.//*[local-name() = "script"]`
* in Symfony: `descendant-or-self::script`

The `|` in the CSS selector is the namespace prefix separator. The Syntax is `prefix|tagname`. Unlike Xpath 1.0, CSS selectors allow the asterisk as a joker for the prefix or the name. So your last CSS selector is broken: http://xpath.thomas.weinert.info/?selector=x%7C.someclass

For the crashes, you could try to increase your error level and/or check the log files. If you send me a document I could try to reproduce it.

**Comment by bkonia** [comment]
That's not what I meant by the "contains" selector. The selector you referenced will find elements in which the specified **attribute** value contains the specified string. I'm referring to the jQuery contains() selector that matches on the string within the element's **text**. For instance:

$fd->find("span:contains('Some Text')")

This works in Symfony. Is there any PHPCss equivalent?

**Comment by ThomasWeinert** [comment]
That is a special jQuery selector, not a CSS 3 selector. No, PHPCss does not implement them yet. I opened a ticket and will think about it.
https://github.com/ThomasWeinert/PhpCss/issues/5

So for now you will have to use the Symfony CSS Selectors and a registered prefix.

Additionally I opened a ticket for the HTML5 loader to allow options for the used parser.
https://github.com/FluentDOM/HTML5/issues/1

I am refactoring the Loader/Serializers integration for the next FluentDOM version to make it more seamless. This will become possible:

```php
$fd = FluentDOM::QueryCss($html, 'text/html5');
$fd->find('html|p')->after('<p>test<br>test</p>');
echo $fd;
```
